core/src/main/scala/org/apache/spark/storage/BlockManagerMasterEndpoint.scala

```diff
@@ -583,7 +586,12 @@ class BlockManagerMasterEndpoint(
 
     val time = System.currentTimeMillis()
     executorIdToLocalDirs.put(id.executorId, localDirs)
-    if (!blockManagerInfo.contains(id)) {
+    // SPARK-41360: For the block manager re-registration, we should only allow it when
+    // the executor is recognized as active by the scheduler backend. Otherwise, this kind
+    // of re-registration from the terminating/stopped executor is meaningless and harmful.
+    lazy val isExecutorAlive =
+      driverEndpoint.askSync[Boolean](CoarseGrainedClusterMessages.IsExecutorAlive(id.executorId))
+    if (!blockManagerInfo.contains(id) && (!isReRegister || isExecutorAlive)) {
```

Comment:
If this is an issue only for terminating executors, we can detect that in executor side and propagate it in the registration request right ?
Else a transient network partition can result in loosing all executors ?

Reply:
> If this is an issue only for terminating executors, we can detect that in executor side and propagate it in the registration request right ? 

What do you mean? The terminating executor can detect itself as being terminating?

Actually, not only the terminating executors,  executors lost due to long GC or executors who failed to be killed by the driver (where the executor could be an orphan rather than terminated ) are also applied here as long as the case is considered executor lost by the driver.

Reply:
`ShutdownHookManager.inShutdown` should tell if it is in the process of shutting down - and prevent call to reregister ?

For the other cases, lost due to long GC, lost due to network partitions, etc - they are legitimate candidates for reregisteration.

Reply:
> ShutdownHookManager.inShutdown should tell if it is in the process of shutting down - and prevent call to reregister ?

Ok, i see.

> For the other cases, lost due to long GC, lost due to network partitions, etc - they are legitimate candidates for registration.

Note that there's a prerequisite of the re-registration in this PR that the executor should already be lost in the driver's view. In that case, block manager re-registration is meaningless since the executor won't reconnect to the driver.

Reply:
Wont the driver not remove in case of a heartbeat expiry even though the executor did not disconnect ? (for the same reasons as above - long gc pause, network partition, etc)

Reply:
> Wont the driver not remove in case of a heartbeat expiry even though the executor did not disconnect ?